### PR TITLE
Improve parser performances

### PR DESCRIPTION
**Pull request by ooflorent**
**What kind of change does this PR introduce?**

refactoring

**Did you add tests for your changes?**

no

**If relevant, link to documentation update:**

n/a

**Summary**

- Reduce polymorphic calls of `Parser` and `BasicEvaluatedExpression`
- Replace some closures by `for of` loops to limit GC

**Does this PR introduce a breaking change?**

no
EDIT (sokra): yes, `BasicExpression.setXXX(null)` is no longer supported

**Other information**

The following changes speed up my builds from 5%.

### CODE REVIEW

**Comment by wa-Nadoo** [comment]
I think you should not mix two types of performance optimizations in one PR.
In my experience, a traversal with 'for of' is always slower than with 'forEach'. IMHO, in this case for best performance, we need to use good old 'for' statement.

**Comment by ooflorent** [comment]
Nowadays forEach, for and for-loops are on par. Moreover changing the loop
does not grant an immediate perf boost but results in less GC pressure. By
profiling webpack builds I noticed that 30% of the build time was due to
GC. This is insane. Removing closures will result in fewer GC cycles, so
better perf.
-- 
Florent Cailhol

**Comment by sokra** [comment]
Note that this PR comes with some major functional changes in BasicEvaluatedExpression:

* i. e. `setNumber(null)` does no longer unset the number.
* BasicEvaluatedExpression can no longer have multiple types.

from the successful CI I assume this is not used in webpack, so I guess it's safe to remove.

**Comment by ooflorent** [comment]
Same goes for unsetWrapped. I was hesitant of removing this but keeping
dead code isn't ideal
-- 
Florent Cailhol

File: lib/Parser.js [hunk 1]
```diff
@@ -1732,9 +1744,8 @@ class Parser extends Tapable {
 
 		const arr = [];
 		if(expression.elements)
-			expression.elements.forEach(expr => {
```
Comment (sokra) [inline lib/Parser.js]: parseStringArray and parseCalculatedStringArray is also unused, could you remove them too?

**Comment by sokra** [comment]
Thanks

**Pull request merged**